plugins/l4d/l4d.cpp

```diff
@@ -3,102 +3,173 @@
 // that can be found in the LICENSE file at the root of the
 // Mumble source tree or at <https://www.mumble.info/LICENSE>.
 
-#include "../mumble_plugin_win32_32bit.h"
+#include "../mumble_plugin_win32_32bit.h" // Include standard plugin header.
+#include "../mumble_plugin_utils.h" // Include plugin header for special functions, like "escape".
 
-procptr32_t posptr, rotptr;
+procptr32_t steamclient, engine; // Variables to contain modules addresses
 
-static bool calcout(float *pos, float *rot, float *opos, float *front, float *top) {
-	float h = rot[0];
-	float v = rot[1];
+static int fetch(float *avatar_pos, float *avatar_front, float *avatar_top, float *camera_pos, float *camera_front, float *camera_top, std::string &context, std::wstring &identity) {
+	for (int i=0;i<3;i++) {
+		avatar_pos[i] = avatar_front[i] = avatar_top[i] = camera_pos[i] = camera_front[i] = camera_top[i] = 0.0f;
+	}
 
-	if ((v < -180.0f) || (v > 180.0f) || (h < -180.0f) || (h > 180.0f))
+	// Boolean value to check if game addresses retrieval is successful
+	bool ok;
+	// Create containers to stuff our raw data into, so we can convert it to Mumble's coordinate system
+	float avatar_pos_corrector[3], camera_pos_corrector[3], camera_front_corrector[3], camera_top_corrector[3];
+	// Char values for extra features
+	char serverid[22], host[22], map[30], playerid[22];
+	// State
+	bool state;
+
+	// State pointers
+	procptr32_t state_base = peekProc<procptr32_t>(pModule + 0x5A7354);
+	if (!state_base) return false;
+	procptr32_t state_offset = peekProc<procptr32_t>(state_base + 0x5C);
+	if (!state_offset) return false;
+
+	// Peekproc and assign game addresses to our containers, so we can retrieve positional data
+	ok = peekProc(state_offset + 0x8, &state, 1) && // Magical state value: 0 when not playing and 1 when in-game.
```

Comment:
For all of these peekProcs:

Please use the short form peekProc: https://github.com/mumble-voip/mumble/blob/master/plugins/mumble_plugin_win32_ptr_type.h#L84

It will automatically use the correct size for the target.

So:

`peekProc(pModule + 0x512264, avatar_pos_corrector, 12)`
becomes
`peekProc(pModule + 0x512264, avatar_pos_corrector)`

etc.